### ISSUE TITLE
Summing DelayedArray and Matrix

### ISSUE THREAD

**Issue by const-ae** [Bioconductor/DelayedArray#60]
Hi Hervé, 
I have recently started to play around with the `DelayedArray` package and it is just great to handle datasets that are too big to be kept in memory.

However, I recently tried to sum a classical `matrix` with a `DelayedArray` of the same dimension. To my surprise this gave the following error:

``` r
suppressPackageStartupMessages({
  library(DelayedArray)
})
# Standard Matrix
mat1 <- matrix(1:12, nrow = 3, ncol = 4)
mat2 <- matrix(3, nrow = 3, ncol = 4)
mat1 + mat2
#>      [,1] [,2] [,3] [,4]
#> [1,]    4    7   10   13
#> [2,]    5    8   11   14
#> [3,]    6    9   12   15

# Operations with DelayedArray
dmat1 <- DelayedArray(mat1)
dmat2 <- DelayedArray(mat2)
dmat1 + dmat2
#> <3 x 4> DelayedMatrix object of type "double":
#>      [,1] [,2] [,3] [,4]
#> [1,]    4    7   10   13
#> [2,]    5    8   11   14
#> [3,]    6    9   12   15

# Mixed
dmat1 + mat2
#> Error in .check_Ops_vector_arg_length(e, x_nrow, e_what = e_what, x_what = x_what): right object is longer than first dimension of left object
```

<sup>Created on 2020-01-23 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0)</sup>

My question is: is this the intended behavior?

I realize that this is not the most realistic scenario because if `dmat1` is too big for the memory, `mat2` would probably be as well. However, I expected this to work because now, if I write a generic function, I will have to check the classes of the input arguments and make sure that they match.

**Comment by hpages** [comment]
Hi Constantin,

The error is expected but the error message is not really helping.

From a technical point of view arithmetic operations between DelayedArray and non-DelayedArray operands are actually only supported when the non-DelayedArray operand is a "small" vector. The typical situations are:
```
x +   5  # add single value
x + 1:2  # add small vector (after recycling it)
```
"small" here means that the length of the vector should not be longer than the 1st dimension of `x`. Another restriction is that its length must be a divisor of the 1st dimension of `x`.

For convenience, I also added support for ordinary matrices or arrays but they're basically treated as vectors (they go thru `as.vector()`). This explains the error message that you get.

I've tried to improve the error message a little bit in DelayedArray 0.13.4 (commit cdfd4d6d):
```
> dmat1 + mat2
Error in .check_Ops_vector_arg_length(e, x_nrow, e_what = e_what, x_what = x_what) : 
  when the right operand is not a DelayedArray object (or derivative),
  its length (12) cannot be greater than the first dimension of the left
  operand (3)
```

H.
